# Ejercicio de manipulación de datos

Para realizar este ejercicio vamos a utilizar el fichero `UCM-Acceso-2024-25-Anonimizado.csv` de la página https://www.universidata.es/datasets/ucm-acceso. Desde dicha ágina podemos acceder al **Diccionario de Datos** 
https://dimetrical.atlassian.net/wiki/spaces/UNC/pages/1608286213/Dataset+Acceso con la descripción de cada campo del fichero.



## Carga de datos y exploración del data frame


Vamos a descargamos los ficheros correspondientes al curso 2024-25 (UCM-Acceso-2024-25-Anonimizado.csv) de la web
https://www.universidata.es/datasets/ucm-acceso.
Lo dejamos en nuestra carpeta llamada `Data` y lo importamos como un data frame llamado `df`.



In [1]:
import pandas as pd

df = pd.read_csv("./Data/ucm-acceso-2024-25-anonimizado.csv")


/var/folders/gz/w9zmh03s61l021wmll3psm_c0000gn/T/ipykernel_1961/1471365357.py:3: DtypeWarning: Columns (35) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("./Data/ucm-acceso-2024-25-anonimizado.csv")


Al hacer la carga de los datos desde el fichero hasta el data frame, me ha salido una advertencia (***DtypeWarning***) de Pandas: una de las columnas del csv contiene datos de tipos diferentes (***Columns (35) have mixed types.***). Esto conviene explorarlo, empezamos mirando las dimensiones, es decir, el número de filas y columnas.

In [4]:
df.shape      # número de filas y columnas

(14019, 78)

Vemos el nombre de las 78 columnas.

In [8]:
df.columns

Index(['IMPORTANTE', 'cod_universidad', 'des_universidad', 'curso_academico',
       'cod_campus', 'des_campus', 'cod_centro', 'des_centro',
       'cod_naturaleza_centro', 'des_naturaleza_centro', 'cod_tipo_estudio',
       'des_tipo_estudio', 'cod_titulacion', 'des_titulacion',
       'cod_comunidad_centro_sec', 'des_comunidad_centro_sec',
       'cod_provincia_centro_sec', 'des_provincia_centro_sec',
       'cod_municipio_centro_sec', 'des_municipio_centro_sec',
       'lat_municipio_centro_sec', 'lon_municipio_centro_sec',
       'cod_naturaleza_centro_sec', 'des_naturaleza_centro_sec',
       'nota_admision', 'cod_pais_fin_estudio_acceso',
       'des_pais_fin_estudio_acceso', 'lat_pais_fin_estudio_acceso',
       'lon_pais_fin_estudio_acceso', 'cod_continente_fin_estudio_acceso',
       'des_continente_fin_estudio_acceso',
       'cod_agregacion_paises_fin_estudio_acceso',
       'des_agregacion_paises_fin_estudio_acceso', 'cod_estudio_acceso',
       'des_estudio_acceso', 'cod_e

Nos interesa la columna 35, para ello preguntamos por el index 35 de la lista que devuelve `df.columns`.

In [11]:
df.columns[35]   

'cod_especialidad_acceso'

Veo el contenido de la columna indexada por 35 y de nombre `cod_especialidad_acceso`.

In [133]:
df[['cod_especialidad_acceso']]

,cod_especialidad_acceso
0,NaN
1,0703
2,NaN
3,NaN
4,NaN
...,...
14014,NaN
14015,NaN
14016,204.0
14017,NaN


In [14]:
type('cod_especialidad_acceso')

str

Al ver `0703` tengo claro que es del tipo `str`porque si fuese un entero, o real, no tendría el `0` delante. Pero esto me hace preguntarme ¿es posible que python haya tomado algún campo como entero y haya perdido un `0`?. Comprobarlo *a mano* en el fichero no tiene sentido pues son 14019 filas.
Lo que debo hacer es una carga de datos segura, para ello leo la descripción de los datos.

https://dimetrical.atlassian.net/wiki/spaces/UNC/pages/1608286213/Dataset+Acceso

Otro dato que me chirría es el `204.0`, miro en la documentaión y este código no existe, ha cambiado a float el código `0204: Humanidades y Ciencias Sociales - Bachillerato LOGSE`. 

La columna `cod_especialidad_acceso` contiene dos representaciones diferentes del mismo código:

- '0703'  texto
- 703.0   número

Por tanto, no podemos dejar que Pandas infiera el tipo, porque mezcla texto y números. hay que leer las columnas `cod_` como `string` durante la carga, normalizarlas y después convertirlas a category.

## Problema: Los códigos como 0703 no deberían tratarse como números.

***Si Pandas intenta interpretar el código 0703 como un número, 0703 se convierte en 703, perdiendo el 0 inicial en la carga.***


### Solución 

Vemos que nuestras variables son

- Categóricas: todas las variables `cod_` y `des_`.
- Numéricas: `lat_`, `lon_`, nota_admision y variables de año.
- Indicadores: las variables `ind_` tienen como valores `SÍ` y `NO`. Las estudio por si puede pasarlas a booleanas.
- IMPORTANTE: variable especial que habría que revisar según su contenido.


¿Cómo puedo ver los distintos valores de las variables ind_ e IMPORTANTE?




In [88]:
lista = df['IMPORTANTE'].unique()
print(lista), print('Número de elementos', len(lista))

lista=df['ind_accede_al_SUE_por_1a_vez'].unique()
print(lista), print('Número de elementos', len(lista))

lista=df['ind_se_matricula_en_primero'].unique()
print(lista), print('Número de elementos', len(lista))

lista=df['ind_curso_adaptacion'].unique()
print(lista), print('Número de elementos', len(lista))

lista=df['anio_fin_estudio_acceso'].unique()
print(lista), print('Número de elementos', len(lista))

lista=df['anio_acceso_SUE'].unique()
print(lista), print('Número de elementos', len(lista))

lista=df['anio_nacimiento'].unique()
print(lista), print('Número de elementos', len(lista))


# En anio_fin_estudio_acceso veo 2008. , compruebo que es un valor real

print(df["anio_fin_estudio_acceso"].value_counts())

['LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓLO INDIVIDUO. PARA INTERPRETAR CORRÉCTAMENTE CONSULTE: https://www.universidata.es/anonimización']
Número de elementos 1
['SÍ' 'NO']
Número de elementos 2
['SÍ' 'NO']
Número de elementos 2
['NO']
Número de elementos 1
[  nan 2008.]
Número de elementos 2
[2024 2023 2022 2021 2020 2019 2017 2004 2018 2015 1975 1987 2012 2002
 2007 2008 2000 1983 1998 2016 1988 2011 1985 2014 1993 1996 2013 1994
 1986 1995 2003 1974 2005 1989 2010 1990 1997 1992 1981 1999 2006 1973
 1977]
Número de elementos 43
[2006 2005 2003 2001 1996 1999 2004 1990 2000 2007 2002 1994 1998 1973
 1988 1993 1980 1991 1992 1997 1982 1978 1995 1977 1974 2008 1987 1975
 1956 1958 1966 1964 1983 1967 1989 1970 1969 1972 1981 1985 1984 1979
 1959 1965 1955 1957 1976 1968 1986 1971 1963 1961 1962 1948 1960 1943]
Número de elementos 56
anio_fin_estudio_acceso
2008.0    1
Name: count, dtype: int64


Cambiar explícitamente el tipo de cada columna que puede generar el conflicto (Por ejemplo: `dtype={"IMPORTANTE": "str"}`) es la solución más adecuada, ya que nos permite controlar cómo interpreta Pandas los datos al leer el archivo.

En nuestro caso podemos cambiar `IMPORTANTE` por `str` o `category`. Si es `str`entonces se tratan a los valores como cadenas de caracters y si es `category` entonces, se tratan a los valores como un conjunto limitado de categorías, aunque sea muy, muy, muy grande, es limitado.

Podría hacer una carga que haga *casting* de la siguiente forma:

In [78]:
df = pd.read_csv(
    "./Data/ucm-acceso-2024-25-anonimizado.csv",
    dtype={"cod_especialidad_acceso": "category", "IMPORTANTE": "str"}
)

Pero son 78 variables. Se puede hacer ad-hoc o necesitamos un mecanismo para hacerla de forma automática y que identifique como:

- Categóricas: todas las variables `cod_` y `des_`.
- float: `lat_`, `lon_` y nota_admision
- int: 'anio_fin_estudio_acceso', 'anio_acceso_SUE', 'anio_nacimiento'.
- Booleanas: las variables `ind_` con valores `SÍ` y `NO`. 
- str: la variable IMPORTANTE.

Pero hay un problema importante con las variables `ind_` porque en el CSV tienen los valores `SÍ` y `NO`, hacer un `dtype = "boolean"` no los convierte utomáticamente a True y False. En este caso, es mejor leer primero las variables `ind_` como `string` y después convertirla.

Hay que hacerlo en dos pasos: 

1. Leer el CSV con los tipos que se pueden aplicar directamente
2. Convertir las ind_ de SÍ/NO a booleanas.



In [124]:
import pandas as pd

fichero = "./Data/ucm-acceso-2024-25-anonimizado.csv"


# --------------------------------------------------
# 1. Obtener los nombres de las columnas
# --------------------------------------------------

columnas = pd.read_csv(fichero, nrows=0).columns


# --------------------------------------------------
# 2. Identificar las columnas categóricas
# --------------------------------------------------

columnas_cat = []

for c in columnas:
    if c.startswith(("cod_", "des_")):
        columnas_cat.append(c)


# --------------------------------------------------
# 3. Identificar las columnas Float64
# --------------------------------------------------

columnas_float = []

for c in columnas:
    if c.startswith(("lat_", "lon_")) or c == "nota_admision":
        columnas_float.append(c)


# --------------------------------------------------
# 4. Variables enteras
# --------------------------------------------------

columnas_int = [
    "anio_fin_estudio_acceso",
    "anio_acceso_SUE",
    "anio_nacimiento"
]


# --------------------------------------------------
# 5. Identificar las variables booleanas
# --------------------------------------------------

columnas_bool = []

for c in columnas:
    if c.startswith("ind_"):
        columnas_bool.append(c)


# --------------------------------------------------
# 6. Definir los tipos de datos para la carga
# --------------------------------------------------

dtype = {}

# Las categorías las cargamos primero como string.
# Después las convertiremos a category.
for c in columnas_cat:
    dtype[c] = "string"    # ¿categórica?

for c in columnas_float:
    dtype[c] = "string"

# Los años se cargan como string porque pueden aparecer
# valores como 2008.
for c in columnas_int:
    dtype[c] = "string"

# SÍ / NO se cargan como string y después se convierten a boolean.
for c in columnas_bool:
    dtype[c] = "string"

dtype["IMPORTANTE"] = "string"


# --------------------------------------------------
# 7. Leer el fichero
# --------------------------------------------------

df = pd.read_csv(
    fichero,
    dtype=dtype
)

# --------------------------------------------------
# 8. Convertir las columnas float
# --------------------------------------------------

for c in columnas_float:
    df[c] = (
        df[c]
        .str.strip()
        .str.replace(",", ".", regex=False)
        .astype("Float64")
    )
    
# --------------------------------------------------
# 8. Convertir los códigos a category
# --------------------------------------------------

for c in columnas_cat:
    df[c] = df[c].astype("category")


# --------------------------------------------------
# 9. Convertir los años a enteros
# --------------------------------------------------

for c in columnas_int:
    df[c] = (
        df[c]
        .str.strip()
        .str.replace(".", "", regex=False)
        .astype("Int64")
    )


# --------------------------------------------------
# 10. Convertir SÍ / NO a boolean
# --------------------------------------------------

for c in columnas_bool:
    df[c] = (
        df[c]
        .str.strip()
        .map({
            "SÍ": True,
            "SI": True,
            "NO": False
        })
        .astype("boolean")
    )


# --------------------------------------------------
# 11. Comprobar el resultado
# --------------------------------------------------

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14019 entries, 0 to 14018
Data columns (total 78 columns):
 #   Column                                    Non-Null Count  Dtype   
---  ------                                    --------------  -----   
 0   IMPORTANTE                                14019 non-null  string  
 1   cod_universidad                           14019 non-null  category
 2   des_universidad                           14019 non-null  category
 3   curso_academico                           14019 non-null  object  
 4   cod_campus                                14019 non-null  category
 5   des_campus                                14019 non-null  category
 6   cod_centro                                13993 non-null  category
 7   des_centro                                13993 non-null  category
 8   cod_naturaleza_centro                     14019 non-null  category
 9   des_naturaleza_centro                     14019 non-null  category
 10  cod_tipo_estudio      

In [100]:
print(columnas[20])

lat_municipio_centro_sec


In [102]:
df['lat_municipio_centro_sec'].unique()

array([40.41669, 42.23136, 38.61524, 40.32642,      nan, 38.5623 ,
       40.54754, 39.50206, 40.54751, 38.37923, 40.4352 , 36.84644,
       40.28387, 39.88757, 43.31584, 40.49286, 41.51669, 28.6104 ,
       38.2655 , 39.46997, 40.68117, 37.83504, 41.65629, 40.22972,
       40.42553, 40.30498, 36.50994, 40.47283, 40.24454, 40.03209,
       39.46688, 40.34912, 40.3297 , 40.62981, 40.56503, 40.0305 ,
       40.48166, 40.57628, 41.61415, 40.32277, 36.59695, 39.85678,
       40.50481, 40.26884, 40.44681, 40.19269, 40.40525, 40.10443,
       40.59907, 40.3571 , 42.33305, 41.93029, 42.46577, 38.68604,
       43.00071, 40.0842 , 40.7035 , 40.30095, 40.59167, 40.58433,
       40.4589 , 39.57144, 36.16118, 40.38393, 43.42711, 42.59988,
       40.15934, 40.75428, 42.34001, 40.14019, 40.65895, 28.46981,
       40.86399, 38.99765, 39.96031, 42.84641, 39.98128, 40.62725,
       40.75522, 40.67348, 40.42594, 42.43362, 39.9598 , 37.76574,
       39.59079, 40.67867, 40.12416, 38.9861 , 41.38792, 40.54

In [106]:
df_prueba = pd.read_csv(fichero)

print(df_prueba.dtypes)
print(df_prueba.iloc[0, 20])


IMPORTANTE                   object
cod_universidad               int64
des_universidad              object
curso_academico              object
cod_campus                    int64
                             ...   
des_ocupacion_madre          object
cod_nivel_estudios_padre    float64
des_nivel_estudios_padre     object
cod_ocupacion_padre         float64
des_ocupacion_padre          object
Length: 78, dtype: object
40.41669


/var/folders/gz/w9zmh03s61l021wmll3psm_c0000gn/T/ipykernel_1961/3841710108.py:1: DtypeWarning: Columns (35) have mixed types. Specify dtype option on import or set low_memory=False.
  df_prueba = pd.read_csv(fichero)


In [118]:
print(df_prueba["cod_especialidad_acceso"].unique())

[nan '0703' '0702' '0204' '0423' '0102' '0421' '0103' '0412' '0401' '0701'
 '0413' '0402' '0406' '0405' '0302' 'EVCH' '0425' '0411' '0417' '0416'
 '0101' '0303' '0410' '0407' '0422' '0802' '0202' '0424' '0301' '0403'
 '0203' '0420' '0404' '0409' '0408' 421.0 702.0 703.0 413.0 423.0 701.0
 420.0 406.0 103.0 408.0 102.0 402.0 101.0 401.0 425.0 409.0 204.0 304.0
 302.0 203.0 410.0 416.0 403.0 422.0 301.0 202.0 412.0 411.0]


¿Quedan más columnas de tipo numérico que necesitan ser categóricas?

Para detectar qué columnas que no son `category`, filtar las no `category` de `Dtype`.

In [155]:
df.loc[:, df.dtypes != "category"].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14019 entries, 0 to 14018
Data columns (total 17 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   IMPORTANTE                    14019 non-null  object 
 1   curso_academico               14019 non-null  object 
 2   lat_municipio_centro_sec      12265 non-null  float64
 3   lon_municipio_centro_sec      12265 non-null  float64
 4   nota_admision                 13365 non-null  object 
 5   lat_pais_fin_estudio_acceso   0 non-null      float64
 6   lon_pais_fin_estudio_acceso   0 non-null      float64
 7   anio_fin_estudio_acceso       1 non-null      float64
 8   ind_accede_al_SUE_por_1a_vez  14019 non-null  object 
 9   anio_acceso_SUE               14019 non-null  int64  
 10  ind_se_matricula_en_primero   14019 non-null  object 
 11  ind_curso_adaptacion          14019 non-null  object 
 12  anio_nacimiento               14019 non-null  int64  
 13  l

Veamos sus datos de uno en uno

In [157]:
df['IMPORTANTE']

0        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
1        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
2        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
3        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
4        LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
                               ...                        
14014    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
14015    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
14016    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
14017    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
14018    LOS DATOS DE ESTA FILA NO CORRESPONDEN A UN SÓ...
Name: IMPORTANTE, Length: 14019, dtype: object

O todos a la vez 

In [164]:
df['ind_accede_al_SUE_por_1a_vez']

0        SÍ
1        SÍ
2        SÍ
3        SÍ
4        SÍ
         ..
14014    SÍ
14015    SÍ
14016    SÍ
14017    NO
14018    SÍ
Name: ind_accede_al_SUE_por_1a_vez, Length: 14019, dtype: object